syncer/syncer.go

```diff
@@ -716,6 +729,58 @@ func (s *Syncer) addCount(isFinished bool, queueBucket string, tp opType, n int6
 	}
 }
 
+func (s *Syncer) jobsIsEmpty() bool {
+	for idx := range s.jobs {
+		if len(s.jobs[idx]) > 0 {
+			return false
+		}
+	}
+	return true
+}
+
+// updateReplicationLag calc syncer replication lag by job, it called after every batch dml job / one skip job / one ddl
+// job is flushed to target db.
+func (s *Syncer) updateReplicationLag(job *job, queueBucketName string) {
+	if job != nil {
+		var lag int64
+		switch job.tp {
+		case ddl:
+			failpoint.Inject("BlockSyncerUpdateDDLLag", func(v failpoint.Value) {
+				t := v.(int)
+				s.tctx.L().Info("BlockSyncerUpdateDDLLag", zap.Int("sleep time", t))
+				time.Sleep(time.Duration(t) * time.Second)
+			})
+			// NOTE: we handle ddl job separately because ddl job will clean all the dml job before execution
+			lag = time.Now().Unix() + s.tsOffset.Load() - int64(job.eventHeader.Timestamp)
+		default: // dml job or skip job
+			if job.tp != skip {
+				failpoint.Inject("BlockSyncerUpdateDMLLag", func(v failpoint.Value) {
+					t := v.(int)
+					s.tctx.L().Info("BlockSyncerUpdateDMLLag", zap.Int("sleep time", t))
+					time.Sleep(time.Duration(t) * time.Second)
+				})
+				// NOTE workerlagmap already init all dml key(queueBucketName) before syncer running.
+				s.workerLagMap[queueBucketName].Store(time.Now().Unix() + s.tsOffset.Load() - int64(job.eventHeader.Timestamp))
+			} else {
+				lag = time.Now().Unix() + s.tsOffset.Load() - int64(job.eventHeader.Timestamp)
```

Comment:
Can we set `s.workerLagMap[queueBucketName]` to 0 when job is nil?
Now if dml jobs are finished, `s.workerLagMap[queueBucketName]` might be a large number. `s.secondsBehindMaster` may be set to it instead of lag here.

Reply:
added